packages/component/src/hooks/internal/useDisableOnBlurEffect.js

```diff
@@ -0,0 +1,30 @@
+import { useEffect } from 'react';
+
+// Because different browser handling disabled-but-focused UI different, we are implementing a more general solution.
+// When the button is being disabled:
+// - If it is not currently focused, we will set "disabled";
+// - If it is currently focused, we will not set "disabled" until it is blurred (focus moved away).
+export default function useDisableOnBlurEffect(targetRef, disabled) {
+  useEffect(() => {
+    if (disabled) {
+      const { current } = targetRef;
+
+      if (current) {
+        if (document.activeElement !== current) {
+          current.setAttribute('disabled', 'disabled');
+
+          return () => current.removeAttribute('disabled');
+        }
+
+        const handler = () => current.setAttribute('disabled', 'disabled');
```

Comment:
Should you move this out of the useEffect hook so you can use useCallback?

Reply:
When I work with `useCallback` or `useMemo`, I always think about our gains. For example, less waste-render, less reconciliation, reduced computation (thru increased memory usage), etc.

If we use `useMemo` with similar memory usage characteristics, the code would become:

```js
export default function useDisableOnBlurEffect(targetRef, disabled) {
  const disabledHandler = useMemo(
    () => disabled && targetRef.current && (() => targetRef.current.setAttribute('disabled', 'disabled')),
    [disabled, targetRef]
  );

  useEffect(() => {
    if (disabled) {
      const { current } = targetRef;

      if (current) {
        if (document.activeElement !== current) {
          current.setAttribute('disabled', 'disabled');

          return () => current.removeAttribute('disabled');
        }

        current.addEventListener('blur', disabledHandler);

        return () => {
          current.removeAttribute('disabled');
          current.removeEventListener('blur', disabledHandler);
        };
      }
    }
  }, [disabled, disabledHandler, targetRef]);
}
```

IMO, it seems a bit over-engineering since the gain is not significant to non-existent, I believe. (See the deps array, they are the same, so I believe no gains.)

If I simplify the code a bit, so it don't look over-engineering:

```js
export default function useDisableOnBlurEffect(targetRef, disabled) {
  const disabledHandler = useMemo(
    () => targetRef.current && (() => targetRef.current.setAttribute('disabled', 'disabled')),
    [targetRef]
  );

  useEffect(() => {
    if (disabled) {
      const { current } = targetRef;

      if (current) {
        if (document.activeElement !== current) {
          current.setAttribute('disabled', 'disabled');

          return () => current.removeAttribute('disabled');
        }

        current.addEventListener('blur', disabledHandler);

        return () => {
          current.removeAttribute('disabled');
          current.removeEventListener('blur', disabledHandler);
        };
      }
    }
  }, [disabled, disabledHandler, targetRef]);
}
```

Then:

- Event handler is always in memory, saving function construction time but increasing heap size
- I believe most widget is 80% enabled, 20% disabled, that means 80% of time we are using more unnecessary memory
- If `disabled` is flipped very frequently, this approach will have some gain on perf, but we aren't flipping it very often

I end up using `bind`:

```
function disableElement(element) {
  element.setAttribute('disabled', 'disabled');
}

export default function useDisableOnBlurEffect(targetRef, disabled) {
  useEffect(() => {
    if (disabled) {
      const { current } = targetRef;

      if (current) {
        if (document.activeElement !== current) {
          current.setAttribute('disabled', 'disabled');

          return () => current.removeAttribute('disabled');
        }

        const blurHandler = disableElement.bind(current);

        current.addEventListener('blur', blurHandler);

        return () => {
          current.removeAttribute('disabled');
          current.removeEventListener('blur', blurHandler);
        };
      }
    }
  }, [disabled, targetRef]);
}
```

I think this would reduce memory usage, and according to [jsperf.com](https://jsperf.com/bind-vs-closure-setup/58), this has same perf compare to closure.